tiatoolbox/tools/registration/wsi_registration.py

```diff
@@ -1059,3 +1064,146 @@ def register(
         image_transform = inverse_translation @ tissue_transform @ forward_translation
 
         return image_transform @ transform_initializer
+
+
+def estimate_bspline_transform(
+    fixed_image: np.ndarray,
+    moving_image: np.ndarray,
+    fixed_mask: np.ndarray,
+    moving_mask: np.ndarray,
+    grid_space: float = 50.0,
+    scale_factors: list = (1, 2, 5),
+    shrink_factor: list = (4, 2, 1),
+    smooth_sigmas: list = (4, 2, 1),
+    num_iterations: int = 100,
+    sampling_percent: float = 0.2,
+):
+    """Estimate BSpline transformation.
+
+    This function performs registration using multi-resolution Bspline approach.
+
+    Args:
+        fixed_image (:class:`numpy.ndarray`):
+            A fixed image.
+        moving_image (:class:`numpy.ndarray`):
+            A moving image.
+        fixed_mask (:class:`numpy.ndarray`):
+            A binary tissue mask for the fixed image.
+        moving_mask (:class:`numpy.ndarray`):
+            A binary tissue mask for the moving image.
+        grid_space:
+            Grid_space (mm) to decide control points.
+        scale_factors:
+            Scaling factor of each BSpline per level in a multi-level setting.
+        shrink_factor:
+            Shrink factor per level to change the size and complexity of the image.
+        smooth_sigmas:
+            Standard deviation for gaussian smoothing per level.
+        num_iterations:
+            Number of iterations.
+        sampling_percent:
+            Fraction of image used for metric evaluation.
+
+    Returns:
+        2D deformation transformation represented by a grid of control points.
+
+    """
+    fixed_image, moving_image = np.squeeze(fixed_image), np.squeeze(moving_image)
+    if len(fixed_image.shape) == 3 or len(moving_image.shape) == 3:
+        raise ValueError("The input images should be grayscale images.")
+
+    # Inverting intensity values
+    fixed_image_inv = 255 - fixed_image
```

Comment:
255 is only for uint8. probably define a function in misc which scales the values between 0 to 1. e.g., for uint8 it will divide by 255, for uint16 by 65535.

Then you can do `1 - fixed_image` here.

Reply:
@John-P do you mean this function?
https://scikit-image.org/docs/stable/api/skimage.util.html#skimage.util.invert

Reply:
You can also do it with cv2 e.g. `cv2.bitwise_not(image)`

Reply:
`cv2.bitwise_not` will only work on binary images. We don't have a binary image here.

Reply:
Ah yes of course, for some reason I was thinking it was a mask.

Reply:
I have used np.invert (https://numpy.org/doc/stable/reference/generated/numpy.invert.html)
https://github.com/TissueImageAnalytics/tiatoolbox/blob/1760d595678752b932649eaebbf677d9814bdd5a/tiatoolbox/tools/registration/wsi_registration.py#L1122-L1124